x-pack/plugins/enterprise_search/server/routes/app_search/app_search_request_handler.test.ts

```diff
@@ -0,0 +1,177 @@
+    );
+
+    AppSearchAPI.shouldHaveBeenCalledWith(
+      'http://localhost:3002/as/credentials/collection?type=indexed&pageIndex=1',
+      {
+        headers: { Authorization: KibanaAuthHeader },
+      }
+    );
+
+    expect(responseMock.ok).toHaveBeenCalledWith({
+      body: appSearchAPIResponseBody,
+    });
+  });
+
+  describe('when an API request fails', () => {
+    it('should return 404 with a message', async () => {
+      const KibanaAuthHeader = 'Basic 123';
+
+      AppSearchAPI.shouldReturnError();
+
+      const requestHandler = createAppSearchRequestHandler({
+        config: mockConfig,
+        log: mockLogger,
+        path: '/as/credentials/collection',
+        hasValidData: (body) => {
+          return (
+            Array.isArray(body?.results) && typeof body?.meta?.page?.total_results === 'number'
+          );
+        },
+      });
+
+      await requestHandler(
+        null,
+        {
+          headers: {
+            authorization: KibanaAuthHeader,
+          },
+        },
+        responseMock
+      );
+
+      AppSearchAPI.shouldHaveBeenCalledWith(`http://localhost:3002/as/credentials/collection?`, {
+        headers: { Authorization: KibanaAuthHeader },
+      });
+
+      expect(responseMock.notFound).toHaveBeenCalledWith({
+        body: 'cannot-connect',
+      });
+    });
+  });
+
+  describe('when `hasValidData` fails', () => {
+    it('should return 404 with a message', async () => {
+      const appSearchAPIResponseBody = {
+        foo: 'bar',
+      };
+      const KibanaAuthHeader = 'Basic 123';
+
+      AppSearchAPI.shouldReturn(appSearchAPIResponseBody);
+
+      const requestHandler = createAppSearchRequestHandler({
+        config: mockConfig,
+        log: mockLogger,
+        path: '/as/credentials/collection',
+        hasValidData: (body) => {
+          return (
+            Array.isArray(body?.results) && typeof body?.meta?.page?.total_results === 'number'
+          );
+        },
+      });
+
+      await requestHandler(
+        null,
+        {
+          headers: {
+            authorization: KibanaAuthHeader,
+          },
+          query: {
+            type: 'indexed',
+            pageIndex: 1,
+          },
+        },
+        responseMock
+      );
+
+      AppSearchAPI.shouldHaveBeenCalledWith(
+        `http://localhost:3002/as/credentials/collection?type=indexed&pageIndex=1`,
+        {
+          headers: { Authorization: KibanaAuthHeader },
+        }
+      );
+
+      expect(responseMock.notFound).toHaveBeenCalledWith({
+        body: 'cannot-connect',
+      });
+    });
+  });
+});
+
+const AppSearchAPI = {
+  shouldHaveBeenCalledWith(expectedUrl: string, expectedParams: object) {
+    expect(fetchMock).toHaveBeenCalledWith(expectedUrl, expectedParams);
+  },
+  shouldReturn(response: object) {
+    fetchMock.mockImplementation(() => {
+      return Promise.resolve(new Response(JSON.stringify(response)));
+    });
+  },
+  shouldReturnError() {
```

Comment:
I'd like to suggest tweaking the names of these helper vars compared to - to me, a `shouldXYZ` function indicates that it contains an assertion/expect() of some kind and throws if its assertion is met. This is doing the opposite - it's setting up a mock, not asserting.

For clarity I think we should rename `shouldReturn` to `mockReturn` and `shouldReturnError` to `mockReturnError`.

Hopefully that makes sense, let me know if not!